includes/class-sensei.php

```diff
@@ -580,27 +575,88 @@ public function install() {
 	} // End install()
 
 	/**
-	 * Check for plugin updates.
+	 * Checks for plugin update tasks and ensures the current version is set.
 	 *
 	 * @since 2.0.0
 	 */
 	public function update() {
-		if ( ! version_compare( $this->version, get_option( 'sensei-version' ), '>' ) ) {
+		$current_version = get_option( 'sensei-version' );
+		$is_new_install  = ! $current_version && ! $this->course_exists();
+		$is_upgrade      = $current_version && version_compare( $this->version, $current_version, '>' );
+
+		// Make sure the current version is up-to-date.
+		if (
+			! $current_version
+			|| $is_upgrade
+		) {
+			$this->register_plugin_version();
+		}
+
+		// Only proceed if we knew the previous version and this was a new install or an upgrade.
+		if ( $current_version && ! $is_new_install && ! $is_upgrade ) {
 			return;
 		}
 
-		// Mark site as having enrolment data from pre-3.0.0.
-		if ( get_option( 'sensei-version' ) && version_compare( '3.0.0-beta.1', get_option( 'sensei-version' ), '>' ) ) {
+		// Mark site as having enrolment data from legacy instances.
+		if (
+			// If the version is known and the previous version was pre-3.0.0.
+			(
+				$is_upgrade
+				&& version_compare( '3.0.0-beta.1', $current_version, '>' )
+			)
+
+			// If there wasn't a current version set and this isn't a new install, double check to make sure there wasn't any enrolment.
+			|| (
+				! $current_version
+				&& ! $is_new_install
+				&& $this->course_progress_exists()
+			)
+		) {
 			update_option( 'sensei_enrolment_legacy', time() );
 		}
 
-		// Run updates.
-		$this->register_plugin_version();
-
 		// Flush rewrite cache.
 		$this->initiate_rewrite_rules_flush();
 	}
 
+	/**
+	 * Helper function to check to see if any courses exists in the database.
+	 *
+	 * @return bool
+	 */
+	private function course_exists() {
+		$course_counts = wp_count_posts( 'course' );
```

Comment:
`wp_count_posts` has a `post_type_exists` check, and the `course` type is not registered at this point, so it returns empty.
This results in $is_new_install being set, and the 'sensei_enrolment_legacy' flag is not created when upgrading from a DB with no 'sensei-version'.

Reply:
Good catch. I switched to a lightweight sampling direct query.
5e7ea25e05fcce58607bc9c3a918a21f6d23e386